File: server/prototypes/rust/hash_map/src/integration.c
#include <stdio.h>
#include <stdint.h>

typedef struct struct_rust_hash_map rust_hash_map_t;

extern rust_hash_map_t *
rust_hash_map_new(void);

extern void
rust_hash_map_free(rust_hash_map_t *);

extern uint32_t
rust_hash_map_set(const rust_hash_map_t *, const char *key, uint32_t value);

extern uint32_t
rust_hash_map_get(const rust_hash_map_t *, const char *key);

int main(void) {}

File: server/prototypes/rust/hash_map/src/integration.rb
require 'ffi'

class Rust < FFI::AutoPointer
  def self.release(ptr)
    HashMap.free(ptr)
  end

  def populate
    HashMap.populate(self)
  end

  def set(key, value)
    HashMap.set(self, key, value)
  end

  def get(key)
    HashMap.get(self, key)
  end

  module HashMap
    extend FFI::Library
    ffi_lib 'integration'

    attach_function :new,  :rust_hash_map_new, [], Rust
    attach_function :free, :rust_hash_map_free, [Rust], :void
                    
    attach_function :get, :rust_hash_map_get, [Rust, :string], :uint32
    attach_function :set, :rust_hash_map_set, [Rust, :string, :uint32], :uint32
  end
end